### ISSUE TITLE
/proc/uid_stat does not exist, change to android.net.TrafficStats API

### ISSUE THREAD

**Issue by x11x** [ukanth/afwall#933]
This is not a bug in Afwall (that I can tell), but I noticed that `/proc/uid_stat` is not present on my Oneplus 6 stock OxygenOS 9.0.2 (based on Android Pie). `uname -r` is `4.9.106-perf+`.
I haven't been able to find any information about what has happened (has it been renamed or deprecated in later versions of Android?). There is not much info on UID-based procfs info as these seem to be Android extensions. E.g. `/proc/uid_*` are not mentioned in the procfs manpage. The manpage is mostly about PID-specific information.
I'm wondering if anyone knows anything about this or has any links to documentation about procfs in Android (mentioning UID-specific info)?
Or, does anyone know of an _alternative way_ to get the information that used to be in `/proc/uid_stat` and can we update the Afwall wiki to describe this? (the wiki currently suggests listing `/proc/uid_stat`)
Also, I noticed that `/proc/uid_stat` is used in the Afwall code but not sure if it being missing will be an issue for Afwall.

**Comment by x11x** [comment]
So I just checked again, `/proc/uid_stat` is used for the App Details listing reporting network statistics. So this will not work with missing `/proc/uid_stat`. If we find a different way to get this info, AppDetailActivity.java could be updated.

**Comment by x11x** [comment]
Some Android filesystem documentation here https://source.android.com/devices/architecture/kernel/reqs-interfaces but I could not find the relevant information and `/proc/uid_stat` is not mentioned.

**Comment by x11x** [comment]
Some Android filesystem documentation here https://source.android.com/devices/architecture/kernel/reqs-interfaces but I could not find the relevant information and `/proc/uid_stat` is not mentioned.

**Comment by x11x** [comment]
Aha I've just found `/proc/net/xt_qtaguid/stat`, looks like it has the information we need although not as easily accessible as `/proc/uid_stat` -- requires parsing the file. It is mentioned here https://stackoverflow.com/a/12939585

**Update**:
Now I've found that [xt_qtaguid](https://source.android.com/devices/tech/datausage/kernel-overview.html) is being deprecated to be replaced by a new system, [eBPF Traffic Monitoring](https://source.android.com/devices/tech/datausage/ebpf-traffic-monitor), starting in Android 9. Interesting slides [here](https://blog.linuxplumbersconf.org/2017/ocw/sessions/4786) and [here](https://blog.linuxplumbersconf.org/2013/ocw//system/presentations/1491/original/Android's%20Netfilter%20Changes%20-%20LPC%2020130920.pdf).
My understanding is that apps should be using the [android.net.TrafficStats](https://developer.android.com/reference/android/net/TrafficStats) API instead of reading the procfs.
My Java is a bit rusty or I would have a go at editing `AppDetailsActivity.java`.

On this same topic, it would be nice to have a page in Afwall listing all apps that have used the network since boot (UIDs, TX and RX numbers), similar to listing `/proc/uid_stat`, for the same purpose of checking if the firewall is working and if there are any boot leaks.

**Comment by ukanth** [comment]
I was planning to rewrite this module with TrafficStats. When I started TrafficStats were in very early stages of development. I'll keep this open. Feel free to explore and make a PR if you are willing to contribute.

**Comment by x11x** [comment]
An update:
I've had a go at it but discovered that TrafficStats is useless. It returns 0 for any UID other than the calling UID. I read somewhere that the app must be signed with a system signature to get TrafficStats info for other UIDs. Simply installing in priv-app does not work.
I also tried requesting permission PACKAGE_USAGE_STATS but that has no effect. (I posted an issue on the Android issue tracker with feature request for this -- to match NetworkStatsManager).

So NetworkStatsManager is probably the way to go as it allows getting usage data for other UIDs when granted PACKAGE_USAGE_STATS permission. NSM methods need a start and end time (e.g. we would have to pass in current time - boot time as the start time if we want the usage data since boot). I'm a bit unclear how this system works and whether it is possible to get accurate per-UID data from it (it returns data back in buckets which are possibly scaled according to your start and end time parameters) -- but I will have a play around with it.

Otherwise, parsing the xt_qtaguid files manually and manually adding support for the new eBPF system when it is implemented (this will have an NDK API according to the docs). I am looking into this too.
